src/node.rs

```diff
@@ -0,0 +1,279 @@
+// Sprengui
+// Copyright © 2020-2021 Jeron Aldaron Lau.
+//
+// Licensed under any of:
+//  - Apache License, Version 2.0 (https://www.apache.org/licenses/LICENSE-2.0)
+//  - MIT License (https://mit-license.org/)
+//  - Boost Software License, Version 1.0 (https://www.boost.org/LICENSE_1_0.txt)
+// At your choosing (See accompanying files LICENSE_APACHE_2_0.txt,
+// LICENSE_MIT.txt and LICENSE_BOOST_1_0.txt).
+
+#![warn(
+    clippy::all,
+    clippy::restriction,
+    clippy::pedantic,
+    clippy::nursery,
+    clippy::cargo
+)]
+#![allow(clippy::blanket_clippy_restriction_lints)]
+#![allow(clippy::missing_inline_in_public_items)]
+#![allow(clippy::module_name_repetitions)]
+#![allow(clippy::exhaustive_structs)]
+#![allow(clippy::implicit_return)]
+
+use crate::script::{NodeScript, Script, SpecialScript};
+use crate::{node, RealEq};
+use std::marker::PhantomData;
+use std::ops::Deref;
+
+/// The 2D section of the node tree
+pub mod node2d {
+    use crate::node::{Node, NodeStruct};
+    use crate::script::{NoScript, NodeScript};
+
+    /// The blank 2d node
+    pub struct Node2D {
+        node: NodeStruct<Self, NodeScript<Self>, NoScript>,
+    }
+
+    impl Node<NodeScript<Self>> for Node2D {
+        fn attach_script(&mut self, script: NodeScript<Self>) {
+            self.node.script = Some(script);
+        }
+
+        fn detach_script(&mut self) -> bool {
+            if let Some(_script) = self.node.script {
+                self.node.script = None;
+                true
+            } else {
+                false
+            }
+        }
+
+        fn node_id(&self) -> u64 {
+            self.node.id
+        }
+    }
+
+    /// All GUI related nodes
+    pub mod gui {
+        /// Button nodes
+        pub mod button {
+            use crate::{
+                node::{Node, NodeStruct},
+                script::{ButtonScript, NodeScript},
+            };
+            use winit::event::WindowEvent;
+            /// Implemented by all button nodes
+            pub trait Button {
+                /// What to do when the button is pressed
+                fn on_click(&mut self, event: &WindowEvent);
+                /// What to do when the button is released
+                fn on_release(&mut self, event: &WindowEvent);
+                /// What to do when the button is hovered over
+                fn on_hover(&mut self);
+            }
```

Comment:
I want sprengui to make use of pasts' [`Notifier`](https://docs.rs/pasts/latest/pasts/trait.Notifier.html) for events.